File: protocol-generator/ocaml/bin/host.ml
(* Standalone peer host — the runnable target for S4 conformance. Boots one
   Peer listener on a TCP port and blocks until signalled, so the entity-core-go
   `validate-peer` oracle can drive the live wire surface against it. Twin of the
   C# `EntityCore.Protocol.Host` and the TS `host.ts`.

     --port N               listen port (default 7777; 0 = auto-assign)
     --debug-open-grants    mint a wide-open admin capability on authenticate
                            instead of the §4.4 restricted standard grant, so
                            the validator can reach grant-gated paths.
     --validate             conformance build (GUIDE-CONFORMANCE §7a): register the
                            system/validate/* test-handlers (echo + dispatch-outbound)
                            so a black-box validator can drive the §6.13(a)/(b) hooks.
                            NOT core protocol; OFF by default (dispatch-outbound is an
                            outbound originator that must never ship live in production).
     --name NAME            load a persistent Ed25519 identity from the standard
                            on-disk location ~/.entity/peers/NAME/keypair (the
                            entity-core PEM keypair: a base64-encoded 32-byte seed
                            between BEGIN/END ENTITY PRIVATE KEY lines — the same
                            convention the Go entity-peer --name and peer-manager use).
                            Without --name a fixed test seed is used (stable peer_id).

   Binds loopback; run validate-peer in the same network namespace. A single
   `LISTENING …` line on stdout signals readiness. *)

open Entitycore_codec

(* Fixed 32-byte Ed25519 seed → stable peer identity across runs (no --name). *)
let default_seed = String.make 32 '\x11'

(* Decode standard-alphabet base64, ignoring whitespace and padding. *)
let b64_decode (s : string) : string =
  let tbl c =
    if c >= 'A' && c <= 'Z' then Char.code c - 65
    else if c >= 'a' && c <= 'z' then Char.code c - 71
    else if c >= '0' && c <= '9' then Char.code c + 4
    else if c = '+' then 62
    else if c = '/' then 63
    else -1
  in
  let buf = Buffer.create (String.length s) and acc = ref 0 and bits = ref 0 in
  String.iter
    (fun c ->
      let v = tbl c in
      if v >= 0 then begin
        acc := (!acc lsl 6) lor v;
        bits := !bits + 6;
        if !bits >= 8 then begin
          bits := !bits - 8;
          Buffer.add_char buf (Char.chr ((!acc lsr !bits) land 0xff))
        end
      end)
    s;
  Buffer.contents buf

(* Load the 32-byte Ed25519 seed from the standard on-disk keypair (§ EXTENSION-
   IDENTITY / Go entity-peer --name): ~/.entity/peers/NAME/keypair, a PEM whose
   body is base64(seed) between BEGIN/END ENTITY PRIVATE KEY lines. *)
let load_seed_from_name (name : string) : string =
  let home = try Sys.getenv "HOME" with Not_found -> "/root" in
  let path =
    List.fold_left Filename.concat home [ ".entity"; "peers"; name; "keypair" ]
  in
  let ic =
    try open_in path
    with Sys_error e -> Printf.eprintf "error: --name %s: %s\n" name e; exit 2
  in
  let lines = ref [] in
  (try while true do lines := input_line ic :: !lines done with End_of_file -> ());
  close_in ic;
  let body =
    String.concat ""
      (List.filter (fun l -> not (String.length l > 0 && l.[0] = '-')) (List.rev !lines))
  in
  let seed = b64_decode (String.trim body) in
  if String.length seed <> 32 then begin
    Printf.eprintf "error: --name %s: expected a 32-byte seed, got %d bytes\n" name
      (String.length seed);
    exit 2
  end;
  seed

let () =
  let port = ref 7777 and open_grants = ref false and validate = ref false in
  let seed = ref default_seed in
  let rec parse = function
    | "--port" :: n :: rest -> port := int_of_string n; parse rest
    | "--name" :: n :: rest -> seed := load_seed_from_name n; parse rest
    | "--debug-open-grants" :: rest ->
        open_grants := true;
        Printf.eprintf
          "warning: --debug-open-grants is DEPRECATED (v7.74 §6.9a; removed v7.75) — it now selects the degenerate `default -> *` seed policy. Prefer --seed-policy.\n%!";
        parse rest
    | "--validate" :: rest -> validate := true; parse rest
    | ("-h" | "--help") :: _ ->
        print_string "usage: host [--port N] [--name NAME] [--debug-open-grants] [--validate]\n"; exit 0
    | [] -> ()
    | arg :: _ -> Printf.eprintf "error: unknown argument '%s'\n" arg; exit 2
  in
  parse (List.tl (Array.to_list Sys.argv));
  let peer = Peer.create ~seed:!seed ~open_grants:!open_grants ~conformance:!validate () in
  let sock, bound = Transport.listen ~port:!port in
  Printf.printf "LISTENING 127.0.0.1:%d peer_id=%s open_grants=%b validate=%b\n%!"
    bound peer.Peer.local_peer !open_grants !validate;
  Transport.accept_loop peer sock

File: protocol-generator/haskell/app/Host.hs
{-# LANGUAGE OverloadedStrings #-}

-- | Standalone peer host — the runnable target for S4 conformance. Boots one
-- 'Peer' listener on a TCP port and blocks, so the entity-core-go @validate-peer@
-- oracle can drive the live wire surface. Twin of the C# @Host@, the TS @host.ts@,
-- and the OCaml @bin/host.ml@.
--
-- @
--   --port N               listen port (default 7777; 0 = auto-assign)
--   --name NAME            load a persistent Ed25519 identity from the standard
--                          on-disk location ~\/.entity\/peers\/NAME\/keypair (the
--                          entity-core PEM keypair: a base64-encoded 32-byte seed
--                          between BEGIN\/END ENTITY PRIVATE KEY lines — the same
--                          convention the Go entity-peer --name and peer-manager
--                          use). Without --name a fixed test seed is used.
--   --debug-open-grants    DEPRECATED (v7.74; removed v7.75): select the
--                          degenerate `default -> *` seed policy.
--   --seed-policy FILE     (next increment; A-HS-011) a seed-policy JSON file.
--   --validate             conformance build (GUIDE-CONFORMANCE §7a): register the
--                          system/validate/* test-handlers. OFF by default.
-- @
module Main (main) where

import Data.Bits (shiftL, shiftR, (.&.), (.|.))
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as BC
import Data.Word (Word8)
import System.Environment (getArgs, lookupEnv)
import System.Exit (exitFailure, exitSuccess, exitWith, ExitCode (ExitFailure))
import System.IO (hPutStrLn, hSetBuffering, stderr, stdout, BufferMode (LineBuffering))
import System.IO.Error (catchIOError)
import qualified Data.Text as T

import EntityCore.Peer (Peer (..), createPeer)
import EntityCore.SeedPolicy (SeedPolicy (..))
import EntityCore.Transport (acceptLoop, listenOn)

-- | Fixed 32-byte Ed25519 seed → stable peer identity across runs (no --name).
defaultSeed :: BC.ByteString
defaultSeed = BC.replicate 32 '\x11'

-- | Decode standard-alphabet base64, ignoring whitespace and padding. Returns
-- 'Nothing' on an invalid character. Hand-rolled (no base64 package is available
-- in the sealed-offline store; depends only on bytestring, already a host dep).
b64Decode :: BC.ByteString -> Maybe B.ByteString
b64Decode = go 0 0 [] . BC.unpack
  where
    val :: Char -> Int
    val c
      | c >= 'A' && c <= 'Z' = fromEnum c - 65
      | c >= 'a' && c <= 'z' = fromEnum c - 71
      | c >= '0' && c <= '9' = fromEnum c + 4
      | c == '+' = 62
      | c == '/' = 63
      | otherwise = -1
    -- acc/bits accumulate sextets; emit a byte once >= 8 bits are buffered.
    go :: Int -> Int -> [Word8] -> [Char] -> Maybe B.ByteString
    go _ _ out [] = Just (B.pack (reverse out))
    go acc bits out (c : rest)
      | c `elem` (" \t\r\n=" :: String) = go acc bits out rest
      | v < 0 = Nothing
      | bits' >= 8 =
          let bits'' = bits' - 8
              byte = fromIntegral ((acc' `shiftR` bits'') .&. 0xff)
          in go acc' bits'' (byte : out) rest
      | otherwise = go acc' bits' out rest
      where
        v = val c
        acc' = (acc `shiftL` 6) .|. v
        bits' = bits + 6

-- | Load the 32-byte Ed25519 seed from the standard on-disk keypair (Go
-- entity-peer --name / peer-manager): ~\/.entity\/peers\/NAME\/keypair, a PEM
-- whose body is base64(seed) between BEGIN\/END ENTITY PRIVATE KEY lines.
loadSeedFromName :: String -> IO BC.ByteString
loadSeedFromName name = do
  home <- maybe "/root" id <$> lookupEnv "HOME"
  let path = home ++ "/.entity/peers/" ++ name ++ "/keypair"
  contents <-
    BC.readFile path `catchIOError` \e -> do
      hPutStrLn stderr ("error: --name " ++ name ++ ": " ++ show e)
      exitWith (ExitFailure 2)
  let body = BC.concat [l | l <- BC.lines contents, not (BC.isPrefixOf "-" l)]
  case b64Decode body of
    Just s | B.length s == 32 -> pure s
    Just s -> do
      hPutStrLn stderr
        ("error: --name " ++ name ++ ": expected a 32-byte seed, got "
          ++ show (B.length s) ++ " bytes")
      exitWith (ExitFailure 2)
    Nothing -> do
      hPutStrLn stderr ("error: --name " ++ name ++ ": malformed base64 keypair body")
      exitWith (ExitFailure 2)

data Opts = Opts {optPort :: Int, optOpen :: Bool, optValidate :: Bool, optSeed :: BC.ByteString}

parse :: [String] -> IO Opts
parse = go (Opts 7777 False False defaultSeed)
  where
    go o [] = pure o
    go o ("--port" : n : rest) = go o {optPort = read n} rest
    go o ("--name" : n : rest) = do
      s <- loadSeedFromName n
      go o {optSeed = s} rest
    go o ("--debug-open-grants" : rest) = do
      hPutStrLn stderr "warning: --debug-open-grants is DEPRECATED (v7.74 §6.9a; removed v7.75) — it now selects the degenerate `default -> *` seed policy. Prefer --seed-policy."
      go o {optOpen = True} rest
    go o ("--validate" : rest) = go o {optValidate = True} rest
    go _ (h : _) | h `elem` ["-h", "--help"] = do
      putStrLn "usage: host [--port N] [--name NAME] [--debug-open-grants] [--validate]"
      exitSuccess
    go _ (arg : _) = do
      hPutStrLn stderr ("error: unknown argument '" ++ arg ++ "'")
      exitFailure

main :: IO ()
main = do
  hSetBuffering stdout LineBuffering
  opts <- getArgs >>= parse
  let policy = if optOpen opts then SeedPolicyDebugOpen else SeedPolicyStandard
  peer <- createPeer (optSeed opts) policy (optValidate opts)
  (sock, bound) <- listenOn (optPort opts)
  putStrLn $
    "LISTENING 127.0.0.1:" ++ show bound
      ++ " peer_id="
      ++ T.unpack (peerLocal peer)
      ++ " open_grants="
      ++ show (optOpen opts)
      ++ " validate="
      ++ show (optValidate opts)
  acceptLoop peer sock
